_drafts/firmware-code-size-2.md

```diff
@@ -0,0 +1,303 @@
+
+If you want to set optimization level for a single function, you can use the `optimize` function attribute:
+```c
+void __attribute__((optimize("O3"))) fast_function(void) {
+    // ...
+}
+```
+
+For a whole file, you can use the `optimize` pragma:
+```c
+#pragma GCC push_options
+#pragma GCC optimize ("O3")
+
+/*
+ * Code that needs optimizing
+ */
+
+#pragma GCC pop_options
+```
+
+Note that the `push_options` and `pop_options` pragmas are needed to make sure the rest of the file is compiled with the regular options.
+
+## Linker Garbage Collection
+Now that we’ve set the appropriate optimization level, we turn our attention to dead code elimination. Most programs contain dead code: it may come from libraries you use partially, or test functions you left in the final program. 
+
+Since the compiler operates on one file at a time, it does not have enough context to decide whether a function is dead code or not.  Consider a library which exposes a function to add numbers in an array: since the files in the library do not call that function, the compiler could think it is not used when other parts of our program rely on it.
+
+The linker on the other hand has visibility into our whole program, so this is where dead code could be identified and removed. Unfortunately, linkers do not perform optimizations by default.
+
+To enable dead code optimization on GCC, you need two things: the compiler needs to split each function into its own linker section so the linker knows where each function is, and the linker needs to add an optimization pass to remove sections that are not called by anything.
+
+This is achieved with the `-ffunction-sections`  compile-time flag, and the `-gc-sections` link-time flag. A similar process can take place with dead data and the `-fdata-sections` flag.
+
+In a Makefile, it looks like this:
+```makefile
+CFLAGS += -ffunction-sections -fdata-sections
+LDFLAGS += -Wl,--gc-sections
+```
+
+ChibiOS sets those when the `USE_LINK_GC` variable is set, so we call:
+```bash
+$ make USE_LINK_GC=yes USE_LTO=no
+```
+
+And get:
+```bash
+Linking build/ch.elf
+Creating build/ch.hex
+Creating build/ch.bin
+Creating build/ch.dmp
+
+   text	   data	    bss	    dec	    hex	filename
+  54828	    228	  98072	 153128	  25628	build/ch.elf
+Creating build/ch.list
+```
+
+So dead code elimination yields close to another 10KiB!
+
+### Preventing some symbols from being garbage collected
+In some cases, you may want some code that is not explicitly called by anything else to remain in your program. One such example is interrupt handlers: while it may look to the linker like they are not called, they the hardware will jump to those addresses and expect the code to be there.
+
+There are two ways to make sure a symbol isn’t garbage collected: with the `used` attribute, or with the `KEEP` linker script command.
+
+Here’s how the `used` attribute is used:
+```c
+void foo(void) __attribute__((used))
```

Comment:
hrm I got the following (which probably applies to the attributes above as well)

```
Compiling file: main.c
../../../main.c:66:1: error: attributes should be specified before the declarator in a function definition
 void foo(void) __attribute__((used)) {
```

I don't think this actually works. It will prevent the compiler from culling the function but its not something the linker looks at ...

Reply:
Removed it.